README.md
# Position

---

[![spm package](http://spmjs.io/badge/position)](http://spmjs.io/package/position)
[![Build Status](https://travis-ci.org/aralejs/position.png)](https://travis-ci.org/aralejs/position)
[![Coverage Status](https://coveralls.io/repos/aralejs/position/badge.png?branch=master)](https://coveralls.io/r/aralejs/position)

简单实用的定位工具，将一个 DOM 节点相对于另一个 DOM 节点进行定位操作。


<span style="font-size:100px;line-height:1;color:#619EE8;">⿻</span>

---

## 安装

```
$ npm install spm -g
$ spm install position
```

## 定位原理

`Position` 组件的定位原理是，通过两个对象分别描述定位元素及其定位点，然后将其定位点重合。

比如有基准元素 A  和 目标元素 B ，先设定 A 的定位点为左上角，B 的定位点为中央。
则分别描述为 `{ element: A, x: 0, y: 0 }` 和 `{ element: B, x: '50%', y: '50%' }`，然后将 B 的中央定位到 A 的左上角，就完成了定位。

定位原理如下图，x 代表横轴，y 代表纵轴。

![定位原理](https://i.alipayobjects.com/e/201307/oOlV3bUVx.png)

## 使用说明

### pin `Position.pin(pinObject, [baseObject])`

基础定位方法，接收两个参数。

1. `pinObject`：目标定位元素，必选。

    类型为字面量对象 `{ element: a, x: 10, y: 10 }`，element 为需定位元素， x 和 y 表示定位元素的定位点。
    
    也可简单写成 DOM 节点 a，相当于 `{ element: a, x: 0, y: 0 }`，表示定位点是节点左上角。
    
2. `baseObject`：基准定位元素，可选。

    类型为字面量对象 `{ element: b, x: 10, y: 10 }`，element 为基准定位元素，x 和 y 表示基准定位元素的定位点。
    
    也可简单写成 DOM 节点 a，相当于 `{ element: a, x: 0, y: 0 }`，表示定位点是节点左上角。
    
    当 element 缺省时，表示 pinObject 相对屏幕可见区域的左上角定位。比如可以写成
        
    ```js
    Position.pin(a, { x: 10, y: 10 });    // 这样后一个参数可简单理解为偏移量
    ```
        
    或写成
        
    ```js
    Position.pin({ element: a, x: -10, y: -10 });
    ```


### center `Position.center(pinElement, [baseElement])`

居中定位，接收两个参数，将 pinElement 定位在 baseElement 元素的中央位置。

1. `pinElement`：定位节点，必选。

2. `baseElement`：基准定位节点，可选。缺省时表示将 pinElement 定位在屏幕中央。


### VIEWPORT `Position.VIEWPORT`

当前可视区域的伪元素，当需要相对于当前可视区域定位时，上述参数的 element 可传入 Position.VIEWPORT。

比如相对于屏幕中央定位：
    
```js
Position.pin(
    { element: a, x: 'center', y: 'center' }, 
    { element: Position.VIEWPORT, x: 'center', y: 'center' }
);
```
    
或写成

```js
Position.center(a, Position.VIEWPORT);
```


## 最佳实践

1. 定位元素到可视区域左上角

    ```js
    Position.pin(a, { x: 0, y: 0 });    // 后一个参数可理解为偏移量
    ```

2. 定位元素到基准元素位置向右偏移 20px

    ```js
    Position.pin(a, { element: b, x: '20px', y: 0 });
    ```

3. 定位元素到基准元素下方 20px 的位置

    ```js
    Position.pin(a, { element: b, x: 0, y: '100%+20px' });
    ```

4. 定位元素到基准元素正中央

    ```js
    Position.center(a, b);
    ```
    
5. 定位元素到基准元素右方中间位置
    
    ```js        
    Position.pin(a, { element: b, x: 'right', y: 'center' });
    ```

    或者
    
    ```js
    Position.pin(a, { element: b, x: '100%', y: '50%' });
    ```

6. 定位元素到可视区域中央

    ```js
    Position.center(a);
    ```


## 注意事项

 - IE8 下，当文档流中定位元素在基准元素的相邻前方，并且不是body的第一个元素时，基准元素有可能会因为未 reflow 而导致 margin-top 失效，所以要尽量避免定位元素与基准元素相邻并在其之前。


HISTORY.md
# 历史记录

---

## 1.1.0

`improved` 升级到 spm@3.x 规范。
`fixed` pinElement 不存在时不再报错。https://github.com/aralejs/position/pull/11

## 1.0.1

`tag:fixed` #6 替换 jquery 1.7.2 中的 offset 方法，修复此版本中判断 body 的 offset 位置的问题。

`tag:improved` #5 替换原来用 $.browser 来判断 ie6 的逻辑。


## 1.0.0

`tag:new` 新组件。


tests/position-spec.js
var expect = require('expect.js');
var Position = require('../index.js');
var $ = require('jquery');

describe('position', function() {

    var pinElement, baseElement, noopDiv;
    
    // reset mocha padding
    $(document.body).css({
        'margin': 0,
        'padding': 0,
        'overflow': 'scroll'
    });

    beforeEach(function() {
        pinElement = $('<div style="width:100px;height:100px;position:absolute;top:0;left:0;" id="pin-element">pinElement</div>').prependTo(document.body);
        // for ie6 bug
        noopDiv = $('<div></div>').prependTo(document.body);
        baseElement = $('<div style="margin:20px;border:5px solid #000;padding:20px;width:200px;height:200px;" id="base-element">baseElement</div>').prependTo(document.body);
    });

    afterEach(function() {
        baseElement.remove();
        noopDiv.remove();
        pinElement.remove();
    });

    it('相对屏幕定位：Position.pin(pinElement, { x: 100, y: 100 })', function() {
        Position.pin(pinElement, { x: 100, y: 100 });
        expect(pinElement.offset().top).to.equal(100);
        expect(pinElement.offset().left).to.equal(100);
    });

    it('基本情况：Position.pin({ element: pinElement, x: 0, y: 0 }, { element:baseElement, x: 100, y: 100 })', function() {
        Position.pin({ element: pinElement, x: 0, y: 0 }, { element:baseElement, x: 100, y: 100 });
        expect(pinElement.offset().top).to.equal(120);
        expect(pinElement.offset().left).to.equal(120);
    });

    it('第一个参数简略写法：Position.pin(pinElement, { element:baseElement, x: 100, y: 100 })', function() {
        Position.pin({ element: pinElement, x: 0, y: 0 }, { element:baseElement, x: 100, y: 100 });
        expect(pinElement.offset().top).to.equal(120);
        expect(pinElement.offset().left).to.equal(120);
    });

    it('第二个参数简略写法：Position.pin({ element:pinElement, x: -100, y: -100 }, baseElement)', function() {
        Position.pin({ element:pinElement, x: -100, y: -100 }, baseElement);
        expect(pinElement.offset().top).to.equal(120);
        expect(pinElement.offset().left).to.equal(120);
    });

    it('两个参数都简略：Position.pin(pinElement, baseElement)', function() {
        Position.pin(pinElement, baseElement);
        expect(pinElement.offset().top).to.equal(20);
        expect(pinElement.offset().left).to.equal(20);
    });

    it('省略x参数：Position.pin(pinElement, { element:baseElement, x: "99px" })', function() {
        Position.pin(pinElement, { element:baseElement, x: "99px" });
        expect(pinElement.offset().top).to.equal(20);
        expect(pinElement.offset().left).to.equal(119);
    });

    it('省略y参数：Position.pin(pinElement, { element:baseElement, y: "99px" })', function() {
        Position.pin(pinElement, { element:baseElement, y: "99px" });
        expect(pinElement.offset().top).to.equal(119);
        expect(pinElement.offset().left).to.equal(20);
    });

    it('带px的字符串参数：Position.pin(pinElement, { element:baseElement, x: "100px", y: "100px" })', function() {
        Position.pin({ element: pinElement, x: 0, y: 0 }, { element:baseElement, x: "100px", y: "100px" });
        expect(pinElement.offset().top).to.equal(120);
        expect(pinElement.offset().left).to.equal(120);
    });

    it('参数传递选择器：Position.pin("#pin-element", { element:"#base-element", x: "100px", y: "100px" })', function() {
        Position.pin('#pin-element', { element:'#base-element', x: "100px", y: "100px" });
        expect(pinElement.offset().top).to.equal(120);
        expect(pinElement.offset().left).to.equal(120);
    });

    it('参数传递原生对象：Position.pin(pinElement[0], { element:baseElement[0], x: "100px", y: "100px" })', function() {
        Position.pin(pinElement[0], { element:baseElement[0], x: "100px", y: "100px" });
        expect(pinElement.offset().top).to.equal(120);
        expect(pinElement.offset().left).to.equal(120);
    });

    it('负数定位点：Position.pin({ element: pinElement, x: -100, y: -100 }, { element:baseElement, x: 0, y: 0 })', function() {
        Position.pin({ element: pinElement, x: -100, y: -100 }, { element:baseElement, x: 0, y: 0 });
        expect(pinElement.offset().top).to.equal(120);
        expect(pinElement.offset().left).to.equal(120);
    });

    it('百分比：Position.pin(pinElement, { element:baseElement, x: "100%", y: "50%" })', function() {
        Position.pin(pinElement, { element:baseElement, x: '100%', y: '50%' });
        expect(pinElement.offset().top).to.equal(145);
        expect(pinElement.offset().left).to.equal(270);
    });

    it('负百分比：Position.pin(pinElement, { element:baseElement, x: "-100%", y: "-50%" })', function() {
        Position.pin(pinElement, { element:baseElement, x: '-100%', y: '-50%' });
        expect(pinElement.offset().top).to.equal(-105);
        expect(pinElement.offset().left).to.equal(-230);
    });

    it('很大的百分比：Position.pin(pinElement, { element:baseElement, x: "1000%", y: "500%" })', function() {
        Position.pin(pinElement, { element:baseElement, x: '1000%', y: '500%' });
        expect(pinElement.offset().top).to.equal(1270);
        expect(pinElement.offset().left).to.equal(2520);
    });

    it('别名：Position.pin({ element:pinElement, x: "left", y: "left" }, { element:baseElement, x: "right", y: "center" })', function() {
        Position.pin({ element:pinElement, x: "left", y: "left" }, { element:baseElement, x: 'right', y: 'center' });
        expect(pinElement.offset().top).to.equal(145);
        expect(pinElement.offset().left).to.equal(270);
    });

    it('大写的别名：Position.pin({ element:pinElement, x: "LEFT", y: "LEFT" }, { element:baseElement, x: "RIGHT", y: "CENTER" })', function() {
        Position.pin({ element:pinElement, x: "LEFT", y: "LEFT" }, { element:baseElement, x: 'RIGHT', y: 'CENTER' });
        expect(pinElement.offset().top).to.equal(145);
        expect(pinElement.offset().left).to.equal(270);
    });

    it('百分比小数：Position.pin(pinElement, { element:baseElement, x: "99.5%", y: "50.5%" })', function() {
        Position.pin(pinElement, { element:baseElement, x: "99.5%", y: "50.5%" });
        expect(pinElement.offset().top).to.be.greaterThan(145.99);
        expect(pinElement.offset().top).to.be.lessThan(147.01);            
        expect(pinElement.offset().left).to.be.greaterThan(267.99);
        expect(pinElement.offset().left).to.be.lessThan(269.01);
    });

    it('居中定位：Position.center(pinElement, baseElement);', function() {
        Position.center(pinElement, baseElement);
        expect(pinElement.offset().top).to.equal(95);
        expect(pinElement.offset().left).to.equal(95);
    });

    it('屏幕居中定位：Position.center(pinElement);', function() {
        Position.center(pinElement);
        expect(($(window).outerHeight()-100)/2).to.be.greaterThan(pinElement.offset().top-0.51);
        expect(($(window).outerHeight()-100)/2).to.be.lessThan(pinElement.offset().top+0.51);
        expect(($(window).outerWidth()-100)/2).to.be.greaterThan(pinElement.offset().left-0.51);
        expect(($(window).outerWidth()-100)/2).to.be.lessThan(pinElement.offset().left+0.51);
    });

    it('baseElement绝对定位：', function() {
        baseElement.css({
            'position': 'absolute',
            'top': '20px',
            'left': '30px'
        });
        Position.pin(pinElement, { element:baseElement, x: 100, y: 100 });
        expect(parseInt(pinElement.offset().top)).to.equal(140);
        expect(parseInt(pinElement.offset().left)).to.equal(150);
    });

    it('baseElement相对定位：', function() {
        baseElement.css({
            'position': 'relative',
            'top': '-20px',
            'left': '30px'
        });
        Position.pin(pinElement, { element:baseElement, x: 100, y: 100 });
        expect(parseInt(pinElement.offset().top)).to.equal(100);
        expect(parseInt(pinElement.offset().left)).to.equal(150);
    });

    it('offsetParent不为body：', function() {
        var offsetParent = $('<div style="margin:20px;border:10px solid #000;padding:20px;position:relative;"></div>').appendTo(document.body);
        baseElement.appendTo(offsetParent);
        Position.pin(pinElement, { element:baseElement, x: 100, y: 100 });
        expect(parseInt(pinElement.offset().top) - parseInt(baseElement.offset().top)).to.equal(100);
        expect(parseInt(pinElement.offset().left) - parseInt(baseElement.offset().left)).to.equal(100);
        offsetParent.remove();
    });

    it('offsetParent绝对定位：', function() {
        var offsetParent = $('<div style="position:absolute;top:50px;left:50px;"></div>').appendTo(document.body);
        baseElement.appendTo(offsetParent);
        Position.pin(pinElement, { element:baseElement, x: 100, y: 100 });
        expect(parseInt(pinElement.offset().top)).to.equal(170);
        expect(parseInt(pinElement.offset().left)).to.equal(170);
        offsetParent.remove();
    });

    it('offsetParent相对定位：', function() {
        var offsetParent = $('<div style="position:relative;top:50px;left:50px;"></div>').prependTo(document.body);
        baseElement.appendTo(offsetParent);
        Position.pin(pinElement, { element:baseElement, x: 100, y: 100 });
        expect(parseInt(pinElement.offset().top)).to.equal(170);
        expect(parseInt(pinElement.offset().left)).to.equal(170);
        offsetParent.remove();
    });

    it('加号应用：', function() {
        Position.pin(pinElement, { element:baseElement, x: "100%+20px", y: "50%+15px" });
        expect(parseInt(pinElement.offset().top)).to.equal(160);
        expect(parseInt(pinElement.offset().left)).to.equal(290);
    });

    it('减号应用：', function() {
        Position.pin(pinElement, { element:baseElement, x: "100%-20px", y: "50%-15px" });
        expect(parseInt(pinElement.offset().top)).to.equal(130);
        expect(parseInt(pinElement.offset().left)).to.equal(250);
    });

    it('加减号混用：', function() {
        Position.pin(pinElement, { element:baseElement, x: "100%-20px+10px", y: "50%-15px+5px" });
        expect(parseInt(pinElement.offset().top)).to.equal(135);
        expect(parseInt(pinElement.offset().left)).to.equal(260);
    });

    it('0% 的情况', function() {
        Position.pin(pinElement, { element:baseElement, x: "0%", y: '0%' });
        expect(parseInt(pinElement.offset().top)).to.equal(20);
        expect(parseInt(pinElement.offset().left)).to.equal(20);
    });

    it('Position.VIEWPORT', function() {
        expect(Position.VIEWPORT._id).to.equal('VIEWPORT');
        expect(Position.VIEWPORT.nodeType).to.equal(1);
    });

    it('Position.VIEWPORT 作为第二个参数（简略）', function() {
        Position.pin(pinElement, Position.VIEWPORT);        
        expect(parseInt(pinElement.offset().top)).to.equal(0);
        expect(parseInt(pinElement.offset().left)).to.equal(0);
    });

    it('Position.VIEWPORT 作为第二个参数（完整）', function() {
        Position.pin(pinElement, {element: Position.VIEWPORT, x: 25, y: 30});        
        expect(parseInt(pinElement.offset().top)).to.equal(30);
        expect(parseInt(pinElement.offset().left)).to.equal(25);
    });

    it('fixed定位：', function() {
        pinElement.css('position', 'fixed');
        Position.pin(pinElement, { x: "300px", y: 250 });
        if ($.browser.msie && $.browser.version == 6.0) {
            expect(pinElement.css('position')).to.equal('absolute');
        } else {
            expect(pinElement.css('position')).to.equal('fixed');
        }
        expect(pinElement.css('top')).to.equal('250px');
        expect(pinElement.css('left')).to.equal('300px');
    });

    it('support 0%+10px', function() {
        Position.pin({ element: pinElement, x: '0%+20px', y: '0%+20px' }, { element:baseElement, x: 100, y: 100 });
        expect(parseInt(pinElement.offset().top)).to.equal(100);
        expect(parseInt(pinElement.offset().left)).to.equal(100);
    });

    it('body 有 margin 的情况', function() {
        $(document.body).css({
            marginTop: 50,
            marginLeft: 50
        });

        Position.pin(pinElement, baseElement);
        expect(pinElement.offset().top).to.equal(baseElement.offset().top);
        expect(pinElement.offset().left).to.equal(baseElement.offset().left);

        $(document.body).css({
            margin: 0
        });
    });

    it('body 是 absolute 的情况', function() {
        $(document.body).css({
            position: 'absolute',
            top: 20,
            left: 20
        });
        Position.pin(pinElement, baseElement);
        expect(parseInt(pinElement.offset().top)).to.equal(40);
        expect(parseInt(pinElement.offset().left)).to.equal(40);

        $(document.body).css({
            position: '',
            top: '',
            left: ''
        });
    });

    it('body 是 relative 的情况', function() {
        $(document.body).css({
            position: 'relative',
            top: 20,
            left: 20
        });

        Position.pin(pinElement, baseElement);
        expect(parseInt(pinElement.offset().top)).to.equal(40);
        expect(parseInt(pinElement.offset().left)).to.equal(40);

        $(document.body).css({
            position: '',
            top: '',
            left: ''
        });
    });

    it("scroll 1000px 之后", function() {
        var temp = $("<div style='height:1000px;'></div>").prependTo(document.body);
        $(document).scrollTop(500);

        Position.pin(pinElement, baseElement);
        expect(parseInt(pinElement.offset().top)>1000).to.be(true);

        temp.remove();
    });

});


src/position.js
define(function(require, exports) {

    // Position
    // --------
    // 定位工具组件，将一个 DOM 节点相对对另一个 DOM 节点进行定位操作。
    // 代码易改，人生难得

    var Position = exports,
        VIEWPORT = { _id: 'VIEWPORT', nodeType: 1 },
        $ = require('$'),
        isPinFixed = false,
        ua = (window.navigator.userAgent || "").toLowerCase(),
        isIE6 = ua.indexOf("msie 6") !== -1;


    // 将目标元素相对于基准元素进行定位
    // 这是 Position 的基础方法，接收两个参数，分别描述了目标元素和基准元素的定位点
    Position.pin = function(pinObject, baseObject) {

        // 将两个参数转换成标准定位对象 { element: a, x: 0, y: 0 }
        pinObject = normalize(pinObject);
        baseObject = normalize(baseObject);

        // if pinObject.element is not present
        var element = pinObject.element;
        if (element === VIEWPORT || element._id === 'VIEWPORT') {
          return;
        }

        // 设定目标元素的 position 为绝对定位
        // 若元素的初始 position 不为 absolute，会影响元素的 display、宽高等属性
        var pinElement = $(element);

        if (pinElement.css('position') !== 'fixed' || isIE6) {
            pinElement.css('position', 'absolute');
            isPinFixed = false;
        }
        else {
            // 定位 fixed 元素的标志位，下面有特殊处理
            isPinFixed = true;
        }

        // 将位置属性归一化为数值
        // 注：必须放在上面这句 `css('position', 'absolute')` 之后，
        //    否则获取的宽高有可能不对
        posConverter(pinObject);
        posConverter(baseObject);

        var parentOffset = getParentOffset(pinElement);
        var baseOffset = baseObject.offset();

        // 计算目标元素的位置
        var top = baseOffset.top + baseObject.y -
                pinObject.y - parentOffset.top;

        var left = baseOffset.left + baseObject.x -
                pinObject.x - parentOffset.left;

        // 定位目标元素
        pinElement.css({ left: left, top: top });
    };


    // 将目标元素相对于基准元素进行居中定位
    // 接受两个参数，分别为目标元素和定位的基准元素，都是 DOM 节点类型
    Position.center = function(pinElement, baseElement) {
        Position.pin({
            element: pinElement,
            x: '50%',
            y: '50%'
        }, {
            element: baseElement,
            x: '50%',
            y: '50%'
        });
    };


    // 这是当前可视区域的伪 DOM 节点
    // 需要相对于当前可视区域定位时，可传入此对象作为 element 参数
    Position.VIEWPORT = VIEWPORT;


    // Helpers
    // -------

    // 将参数包装成标准的定位对象，形似 { element: a, x: 0, y: 0 }
    function normalize(posObject) {
        posObject = toElement(posObject) || {};

        if (posObject.nodeType) {
            posObject = { element: posObject };
        }

        var element = toElement(posObject.element) || VIEWPORT;
        if (element.nodeType !== 1) {
            throw new Error('posObject.element is invalid.');
        }

        var result = {
            element: element,
            x: posObject.x || 0,
            y: posObject.y || 0
        };

        // config 的深度克隆会替换掉 Position.VIEWPORT, 导致直接比较为 false
        var isVIEWPORT = (element === VIEWPORT || element._id === 'VIEWPORT');

        // 归一化 offset
        result.offset = function() {
            // 若定位 fixed 元素，则父元素的 offset 没有意义
            if (isPinFixed) {
                return {
                    left: 0,
                    top: 0
                };
            }
            else if (isVIEWPORT) {
                return {
                    left: $(document).scrollLeft(),
                    top: $(document).scrollTop()
                };
            }
            else {
                return getOffset($(element)[0]);
            }
        };

        // 归一化 size, 含 padding 和 border
        result.size = function() {
            var el = isVIEWPORT ? $(window) : $(element);
            return {
                width: el.outerWidth(),
                height: el.outerHeight()
            };
        };

        return result;
    }

    // 对 x, y 两个参数为 left|center|right|%|px 时的处理，全部处理为纯数字
    function posConverter(pinObject) {
        pinObject.x = xyConverter(pinObject.x, pinObject, 'width');
        pinObject.y = xyConverter(pinObject.y, pinObject, 'height');
    }

    // 处理 x, y 值，都转化为数字
    function xyConverter(x, pinObject, type) {
        // 先转成字符串再说！好处理
        x = x + '';

        // 处理 px
        x = x.replace(/px/gi, '');

        // 处理 alias
        if (/\D/.test(x)) {
            x = x.replace(/(?:top|left)/gi, '0%')
                 .replace(/center/gi, '50%')
                 .replace(/(?:bottom|right)/gi, '100%');
        }

        // 将百分比转为像素值
        if (x.indexOf('%') !== -1) {
            //支持小数
            x = x.replace(/(\d+(?:\.\d+)?)%/gi, function(m, d) {
                return pinObject.size()[type] * (d / 100.0);
            });
        }

        // 处理类似 100%+20px 的情况
        if (/[+\-*\/]/.test(x)) {
            try {
                // eval 会影响压缩
                // new Function 方法效率高于 for 循环拆字符串的方法
                // 参照：http://jsperf.com/eval-newfunction-for
                x = (new Function('return ' + x))();
            } catch (e) {
                throw new Error('Invalid position value: ' + x);
            }
        }

        // 转回为数字
        return numberize(x);
    }

    // 获取 offsetParent 的位置
    function getParentOffset(element) {
        var parent = element.offsetParent();

        // IE7 下，body 子节点的 offsetParent 为 html 元素，其 offset 为
        // { top: 2, left: 2 }，会导致定位差 2 像素，所以这里将 parent
        // 转为 document.body
        if (parent[0] === document.documentElement) {
            parent = $(document.body);
        }

        // 修正 ie6 下 absolute 定位不准的 bug
        if (isIE6) {
            parent.css('zoom', 1);
        }

        // 获取 offsetParent 的 offset
        var offset;

        // 当 offsetParent 为 body，
        // 而且 body 的 position 是 static 时
        // 元素并不按照 body 来定位，而是按 document 定位
        // http://jsfiddle.net/afc163/hN9Tc/2/
        // 因此这里的偏移值直接设为 0 0
        if (parent[0] === document.body &&
            parent.css('position') === 'static') {
                offset = { top:0, left: 0 };
        } else {
            offset = getOffset(parent[0]);
        }

        // 根据基准元素 offsetParent 的 border 宽度，来修正 offsetParent 的基准位置
        offset.top += numberize(parent.css('border-top-width'));
        offset.left += numberize(parent.css('border-left-width'));

        return offset;
    }

    function numberize(s) {
        return parseFloat(s, 10) || 0;
    }

    function toElement(element) {
        return $(element)[0];
    }

    // fix jQuery 1.7.2 offset
    // document.body 的 position 是 absolute 或 relative 时
    // jQuery.offset 方法无法正确获取 body 的偏移值
    //   -> http://jsfiddle.net/afc163/gMAcp/
    // jQuery 1.9.1 已经修正了这个问题
    //   -> http://jsfiddle.net/afc163/gMAcp/1/
    // 这里先实现一份
    // 参照 kissy 和 jquery 1.9.1
    //   -> https://github.com/kissyteam/kissy/blob/master/src/dom/sub-modules/base/src/base/offset.js#L366
    //   -> https://github.com/jquery/jquery/blob/1.9.1/src/offset.js#L28
    function getOffset(element) {
        var box = element.getBoundingClientRect(),
            docElem = document.documentElement;

        // < ie8 不支持 win.pageXOffset, 则使用 docElem.scrollLeft
        return {
            left: box.left + (window.pageXOffset || docElem.scrollLeft) -
                  (docElem.clientLeft || document.body.clientLeft  || 0),
            top: box.top  + (window.pageYOffset || docElem.scrollTop) -
                 (docElem.clientTop || document.body.clientTop  || 0)
        };
    }

});


index.js
// Position
// --------
// 定位工具组件，将一个 DOM 节点相对对另一个 DOM 节点进行定位操作。
// 代码易改，人生难得

var Position = exports,
    VIEWPORT = { _id: 'VIEWPORT', nodeType: 1 },
    $ = require('jquery'),
    isPinFixed = false,
    ua = (window.navigator.userAgent || "").toLowerCase(),
    isIE6 = ua.indexOf("msie 6") !== -1;


// 将目标元素相对于基准元素进行定位
// 这是 Position 的基础方法，接收两个参数，分别描述了目标元素和基准元素的定位点
Position.pin = function(pinObject, baseObject) {

    // 将两个参数转换成标准定位对象 { element: a, x: 0, y: 0 }
    pinObject = normalize(pinObject);
    baseObject = normalize(baseObject);

    // if pinObject.element is not present
    // https://github.com/aralejs/position/pull/11
    if (pinObject.element === VIEWPORT ||
        pinObject.element._id === 'VIEWPORT') {
        return;
    }

    // 设定目标元素的 position 为绝对定位
    // 若元素的初始 position 不为 absolute，会影响元素的 display、宽高等属性
    var pinElement = $(pinObject.element);

    if (pinElement.css('position') !== 'fixed' || isIE6) {
        pinElement.css('position', 'absolute');
        isPinFixed = false;
    }
    else {
        // 定位 fixed 元素的标志位，下面有特殊处理
        isPinFixed = true;
    }

    // 将位置属性归一化为数值
    // 注：必须放在上面这句 `css('position', 'absolute')` 之后，
    //    否则获取的宽高有可能不对
    posConverter(pinObject);
    posConverter(baseObject);

    var parentOffset = getParentOffset(pinElement);
    var baseOffset = baseObject.offset();

    // 计算目标元素的位置
    var top = baseOffset.top + baseObject.y -
            pinObject.y - parentOffset.top;

    var left = baseOffset.left + baseObject.x -
            pinObject.x - parentOffset.left;

    // 定位目标元素
    pinElement.css({ left: left, top: top });
};


// 将目标元素相对于基准元素进行居中定位
// 接受两个参数，分别为目标元素和定位的基准元素，都是 DOM 节点类型
Position.center = function(pinElement, baseElement) {
    Position.pin({
        element: pinElement,
        x: '50%',
        y: '50%'
    }, {
        element: baseElement,
        x: '50%',
        y: '50%'
    });
};


// 这是当前可视区域的伪 DOM 节点
// 需要相对于当前可视区域定位时，可传入此对象作为 element 参数
Position.VIEWPORT = VIEWPORT;


// Helpers
// -------

// 将参数包装成标准的定位对象，形似 { element: a, x: 0, y: 0 }
function normalize(posObject) {
    posObject = toElement(posObject) || {};

    if (posObject.nodeType) {
        posObject = { element: posObject };
    }

    var element = toElement(posObject.element) || VIEWPORT;
    if (element.nodeType !== 1) {
        throw new Error('posObject.element is invalid.');
    }

    var result = {
        element: element,
        x: posObject.x || 0,
        y: posObject.y || 0
    };

    // config 的深度克隆会替换掉 Position.VIEWPORT, 导致直接比较为 false
    var isVIEWPORT = (element === VIEWPORT || element._id === 'VIEWPORT');

    // 归一化 offset
    result.offset = function() {
        // 若定位 fixed 元素，则父元素的 offset 没有意义
        if (isPinFixed) {
            return {
                left: 0,
                top: 0
            };
        }
        else if (isVIEWPORT) {
            return {
                left: $(document).scrollLeft(),
                top: $(document).scrollTop()
            };
        }
        else {
            return getOffset($(element)[0]);
        }
    };

    // 归一化 size, 含 padding 和 border
    result.size = function() {
        var el = isVIEWPORT ? $(window) : $(element);
        return {
            width: el.outerWidth(),
            height: el.outerHeight()
        };
    };

    return result;
}

// 对 x, y 两个参数为 left|center|right|%|px 时的处理，全部处理为纯数字
function posConverter(pinObject) {
    pinObject.x = xyConverter(pinObject.x, pinObject, 'width');
    pinObject.y = xyConverter(pinObject.y, pinObject, 'height');
}

// 处理 x, y 值，都转化为数字
function xyConverter(x, pinObject, type) {
    // 先转成字符串再说！好处理
    x = x + '';

    // 处理 px
    x = x.replace(/px/gi, '');

    // 处理 alias
    if (/\D/.test(x)) {
        x = x.replace(/(?:top|left)/gi, '0%')
             .replace(/center/gi, '50%')
             .replace(/(?:bottom|right)/gi, '100%');
    }

    // 将百分比转为像素值
    if (x.indexOf('%') !== -1) {
        //支持小数
        x = x.replace(/(\d+(?:\.\d+)?)%/gi, function(m, d) {
            return pinObject.size()[type] * (d / 100.0);
        });
    }

    // 处理类似 100%+20px 的情况
    if (/[+\-*\/]/.test(x)) {
        try {
            // eval 会影响压缩
            // new Function 方法效率高于 for 循环拆字符串的方法
            // 参照：http://jsperf.com/eval-newfunction-for
            x = (new Function('return ' + x))();
        } catch (e) {
            throw new Error('Invalid position value: ' + x);
        }
    }

    // 转回为数字
    return numberize(x);
}

// 获取 offsetParent 的位置
function getParentOffset(element) {
    var parent = element.offsetParent();

    // IE7 下，body 子节点的 offsetParent 为 html 元素，其 offset 为
    // { top: 2, left: 2 }，会导致定位差 2 像素，所以这里将 parent
    // 转为 document.body
    if (parent[0] === document.documentElement) {
        parent = $(document.body);
    }

    // 修正 ie6 下 absolute 定位不准的 bug
    if (isIE6) {
        parent.css('zoom', 1);
    }

    // 获取 offsetParent 的 offset
    var offset;

    // 当 offsetParent 为 body，
    // 而且 body 的 position 是 static 时
    // 元素并不按照 body 来定位，而是按 document 定位
    // http://jsfiddle.net/afc163/hN9Tc/2/
    // 因此这里的偏移值直接设为 0 0
    if (parent[0] === document.body &&
        parent.css('position') === 'static') {
            offset = { top:0, left: 0 };
    } else {
        offset = getOffset(parent[0]);
    }

    // 根据基准元素 offsetParent 的 border 宽度，来修正 offsetParent 的基准位置
    offset.top += numberize(parent.css('border-top-width'));
    offset.left += numberize(parent.css('border-left-width'));

    return offset;
}

function numberize(s) {
    return parseFloat(s, 10) || 0;
}

function toElement(element) {
    return $(element)[0];
}

// fix jQuery 1.7.2 offset
// document.body 的 position 是 absolute 或 relative 时
// jQuery.offset 方法无法正确获取 body 的偏移值
//   -> http://jsfiddle.net/afc163/gMAcp/
// jQuery 1.9.1 已经修正了这个问题
//   -> http://jsfiddle.net/afc163/gMAcp/1/
// 这里先实现一份
// 参照 kissy 和 jquery 1.9.1
//   -> https://github.com/kissyteam/kissy/blob/master/src/dom/sub-modules/base/src/base/offset.js#L366
//   -> https://github.com/jquery/jquery/blob/1.9.1/src/offset.js#L28
function getOffset(element) {
    var box = element.getBoundingClientRect(),
        docElem = document.documentElement;

    // < ie8 不支持 win.pageXOffset, 则使用 docElem.scrollLeft
    return {
        left: box.left + (window.pageXOffset || docElem.scrollLeft) -
              (docElem.clientLeft || document.body.clientLeft  || 0),
        top: box.top  + (window.pageYOffset || docElem.scrollTop) -
             (docElem.clientTop || document.body.clientTop  || 0)
    };
}
